### ISSUE TITLE
ArrayIndexOutOfBoundsException in XorCsrfTokenRequestAttributeHandler

### ISSUE THREAD

**Issue by mhankus** [spring-projects/spring-security#13310]
java.lang.ArrayIndexOutOfBoundsException is thrown in XorCsrfTokenRequestAttributeHandler during attack
Affects version spring-security 6.0.3 (I have not tested 6.1)

```
java.lang.ArrayIndexOutOfBoundsException: arraycopy: last destination index 36 out of bounds for byte[8]
        at org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler.xorCsrf(XorCsrfTokenRequestAttributeHandler.java:119) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler.getTokenValue(XorCsrfTokenRequestAttributeHandler.java:99) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler.resolveCsrfTokenValue(XorCsrfTokenRequestAttributeHandler.java:73) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:121) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.9.jar!/:6.0.9]
        at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:185) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:172) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:133) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:90) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:75) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.9.jar!/:6.0.9]
        at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:185) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:172) ~[spring-security-web-6.0.3.jar!/:6.0.3]
        at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:133) ~[spring-security-web-6.0.3.jar!/:6.0.3]

```

To reproduce modify csrf token values on client side (cookie based tokens)

**Expected behavior**
getTokenValue should validate encoded token length and return null if value is incorrect. Generating stacktrace for exception is much more expensive and may impact performance.

**Comment by kevin2jordan** [comment]
Can I work on this issue?

**Comment by jzheaux** [comment]
@kevin2jordan, are you still interested in this issue? A PR would be most welcome!

**Comment by kevin2jordan** [comment]
@jzheaux Have submitted the 
PR: https://github.com/spring-projects/spring-security/pull/13550
Would appreciate the PR review and any comments/feedback for the PR.

**Comment by maximilianschweitzer** [comment]
This issue is not completely fixed yet. During an attack we still get

```
java.lang.ArrayIndexOutOfBoundsException: arraycopy: last destination index 36 out of bounds for byte[1]
	at org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler.xorCsrf(XorCsrfTokenRequestAttributeHandler.java:122)
	at org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler.getTokenValue(XorCsrfTokenRequestAttributeHandler.java:99)
	at org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler.resolveCsrfTokenValue(XorCsrfTokenRequestAttributeHandler.java:73)
	at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:121)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
	at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:91)
```

in version spring-security 6.2.1.
	
Specifically, the problem lies in the following line:

```
System.arraycopy(csrfBytes, 0, xoredCsrf, 0, csrfBytes.length);
```

Here, you're attempting to copy _csrfBytes.length_ elements from _csrfBytes_ into _xoredCsrf_. However, _xoredCsrf_ is only as long as the smaller of the two arrays (_randomBytes.length_ or _csrfBytes.length_), as determined by the preceding line:

```
byte[] xoredCsrf = new byte[len]; // len is the smaller of randomBytes.length or csrfBytes.length
```

This will cause an _ArrayIndexOutOfBoundsException_ when _csrfBytes.length_ is greater than _len_, because you're trying to copy more elements than _xoredCsrf_ can hold.

To fix this, you should copy only _len_ elements, not _csrfBytes.length_ elements, since _len_ is the length of the destination array _xoredCsrf_.

```
// Copy only len elements
System.arraycopy(csrfBytes, 0, xoredCsrf, 0, len);
```

**Comment by pyyx** [comment]
```
java.lang.ArrayIndexOutOfBoundsException: arraycopy: last destination index 36 out of bounds for byte[0]
	at java.lang.System.arraycopy(Native Method) ~[?:?]
	at org.springframework.security.web.csrf.XorCsrfTokenRequestAttributeHandler.xorCsrf(XorCsrfTokenRequestAttributeHandler.java:119) ~[spring-security-web-6.1.8.jar:6.1.8]
```
In version 6.1.8, I had the same problem as @maximilianschweitzer .Hopefully this will be resolved

**Comment by kratosmy** [comment]
Hi, @jzheaux, I have made a quick fix for this issue. Please kindly help review.

**Comment by sjohnr** [comment]
@maximilianschweitzer thanks for reporting the additional finding. I am considering this a bug and plan to backport the fix OSS supported branches (`5.8.x`, `6.2.x` and `6.3.x`) as well. See gh-15184.

@kratosmy thanks for submitting a PR. I will provide feedback on the PR.
